### PR TITLE
Adds interactive mode for info command

### PR DESCRIPTION
**Pull request by PuneetPunamiya**
- This patch interactively ask the users to select

	- Catalog first
   	- Resources in the catalog selected
	- All Versions of the resource selected

  - If there is only one catalog or version then
    it will be auto-selected

  - User can also get the information of the resource
    if the user know the catalog name, resource name
    and version of the resource by

    `tkn hub info task <task-name> --version <version-number> --from catalog <catalog-name>`

Signed-off-by: Puneet Punamiya <ppunamiy@redhat.com>

demo

# Changes

# Submitter Checklist

These are the criteria that every PR should meet, please check them off as you
review them:

- [ ] Includes [tests](https://github.com/tektoncd/community/blob/main/standards.md#principles) (if functionality changed/added)
- [ ] Commit messages follow [commit message best practices](https://github.com/tektoncd/community/blob/main/standards.md#commit-messages)

_See [the contribution guide](https://github.com/tektoncd/pipeline/blob/main/CONTRIBUTING.md) for more details._

### CODE REVIEW

File: api/pkg/cli/cmd/info/info.go [hunk 1]
```diff
@@ -139,8 +139,68 @@ func (opts *options) run() error {
 	}
 
 	hubClient := opts.cli.Hub()
+
+	// Get all Catalogs
+	catalog, err := hubClient.GetCatalogsList()
```
Comment (sm43) [inline api/pkg/cli/cmd/info/info.go:144]: this can be done when user has not passed catalog
```
if opt.from == "" {
  get all catalogs
}
```

File: api/pkg/cli/cmd/info/info.go [hunk 2]
```diff
@@ -139,8 +139,68 @@ func (opts *options) run() error {
...
+		// Ask the catalog name
+		err = selectOption.Ask("catalog", catalog)
+		if err != nil {
+			return err
+		}
+		opts.from = selectOption.Catalog
+	}
+
+	// Get all resources from the Catalog selected
+	resources, err := hubClient.GetResourcesList(hub.SearchOption{
```
Comment (sm43) [inline api/pkg/cli/cmd/info/info.go:163]: this can be done only when user hasn't passed resource name or catalog

Comment (sm43) [reply api/pkg/cli/cmd/info/info.go:163]: if user has passed resource name then this call can be avoided

File: api/pkg/cli/cmd/info/info.go [hunk 3]
```diff
@@ -139,8 +139,68 @@ func (opts *options) run() error {
...
+		// Ask the resource name
+		err = selectOption.Ask(opts.kind, resources)
+		if err != nil {
+			return err
+		}
+		name = selectOption.Name
+	}
+
+	// Get all the versions of the resource selected
+	ver, err := hubClient.GetResourceVersionslist(hub.ResourceOption{
```
Comment (sm43) [inline api/pkg/cli/cmd/info/info.go:182]: similarly here also

File: api/go.mod [hunk 4]
```diff
@@ -14,10 +16,11 @@ require (
 	github.com/joho/godotenv v1.3.0
 	github.com/mitchellh/go-homedir v1.1.0
 	github.com/sergi/go-diff v1.2.0 // indirect
-	github.com/spf13/cobra v1.1.1
+	github.com/spf13/cobra v1.1.3
 	github.com/spf13/viper v1.7.0
 	github.com/stretchr/testify v1.6.1
-	github.com/tektoncd/pipeline v0.20.1-0.20210204110343-8c5a751b53ea
+	github.com/tektoncd/cli v0.18.0
```
Comment (sm43) [inline api/go.mod:22]: not sure if we should import cli dependency .. i think we didn't do this while fetching version
@vinamra28 ?

File: api/pkg/cli/cmd/info/info.go [hunk 5]
```diff
@@ -104,7 +105,7 @@ func Command(cli app.CLI) *cobra.Command {
 		commandForKind("task", opts),
 	)
 
-	cmd.PersistentFlags().StringVar(&opts.from, "from", "tekton", "Name of Catalog to which resource belongs.")
+	cmd.PersistentFlags().StringVar(&opts.from, "from", "", "Name of Catalog to which resource belongs.")
```
Comment (sm43) [inline api/pkg/cli/cmd/info/info.go:108]: `Name of Catalog to which resource belongs. (Default: tekton)`

File: api/pkg/cli/hub/get_catalog.go [hunk 6]
```diff
@@ -0,0 +1,76 @@
+package hub
+
+import (
+	"encoding/json"
+	"net/http"
+
+	cclient "github.com/tektoncd/hub/api/v1/gen/http/catalog/client"
+)
+
+type CatalogOptions struct {
```
Comment (sm43) [inline api/pkg/cli/hub/get_catalog.go:10]: it this used? seems empty

File: api/pkg/cli/hub/get_catalog.go [hunk 7]
```diff
@@ -0,0 +1,76 @@
...
+	data    []byte
+	status  int
+	err     error
+	Catalog CatalogData
+}
+type CatalogData = cclient.ListResponseBody
+
+func (c *client) GetAllCatalogs() CatalogResult {
+	data, status, err := c.Get(catalogEndpoint())
+	if status == http.StatusNotFound {
```
Comment (sm43) [inline api/pkg/cli/hub/get_catalog.go:23]: Api doesn't return 404 currently if there are not catalogs in db... i think it should 🤔

File: api/pkg/cli/hub/get_catalog.go [hunk 8]
```diff
@@ -0,0 +1,76 @@
...
+	c := h.GetAllCatalogs()
+
+	// Unmarshal the data
+	var err error
+	typed, err := c.Type()
+	if err != nil {
+		return nil, err
+	}
+
+	var data = struct {
```
Comment (sm43) [inline api/pkg/cli/hub/get_catalog.go:62]: why is this required? can't we direct loop on CatalogData.Data?

File: api/pkg/cli/hub/get_resource.go [hunk 9]
```diff
@@ -186,3 +187,59 @@ func (rr *ResourceResult) MinPipelinesVersion() (string, error) {
...
+	var resources []string
+	for i := range data.Resources {
+		resources = append(resources, *data.Resources[i].Name)
+	}
+
+	return resources, nil
+}
+
+func (hubClient *client) GetResourceVersionslist(r ResourceOption) ([]string, error) {
+	type options struct {
```
Comment (sm43) [inline api/pkg/cli/hub/get_resource.go:219]: this struct can be moved outside

File: api/go.mod (hunk 4 above)
Comment (vinamra28) [reply api/go.mod:22]: right. This will add cyclic dependency problem

File: api/pkg/cli/cmd/info/info.go (hunk 2 above)
Comment (PuneetPunamiya) [reply api/pkg/cli/cmd/info/info.go:163]: Yeah that makes sense 👍🏻

File: api/pkg/cli/cmd/info/info.go (hunk 3 above)
Comment (PuneetPunamiya) [reply api/pkg/cli/cmd/info/info.go:182]: Yes 👍🏻

File: api/go.mod (hunk 4 above)
Comment (PuneetPunamiya) [reply api/go.mod:22]: Yep will update the change 👍🏻

File: api/pkg/cli/cmd/info/info.go (hunk 1 above)
Comment (PuneetPunamiya) [reply api/pkg/cli/cmd/info/info.go:144]: I think in order to keep default catalog as `tekton` as mentioned [here](https://github.com/tektoncd/hub/pull/265#discussion_r637683325) we'll have to make an api call to get all catalogs

Comment (sm43) [reply api/pkg/cli/cmd/info/info.go:144]: you can keep the `from` flag empty but if user passes a catalog through the flag then you don't need to `get all catalogs`.
use that catalog to fetch the resource, whether it exist or not.

Comment (PuneetPunamiya) [reply api/pkg/cli/cmd/info/info.go]: Yes make sense to get all the catalog if we keep the flag empty

Comment (PuneetPunamiya) [reply api/pkg/cli/cmd/info/info.go]: Updated the PR!

File: api/pkg/cli/cmd/info/info.go [hunk 10]
```diff
@@ -139,8 +139,73 @@ func (opts *options) run() error {
...
+
+	if opts.from == "" {
+		// Get all Catalogs
+		catalog, err := hubClient.GetCatalogsList()
+		if err != nil {
+			return err
+		}
+		if err != nil {
+			return err
+		}
```
Comment (sm43) [inline api/pkg/cli/cmd/info/info.go:153]: added twice

Comment (PuneetPunamiya) [reply api/pkg/cli/cmd/info/info.go]: Updated !

File: api/pkg/cli/cmd/info/info.go [hunk 11]
```diff
@@ -139,8 +139,70 @@ func (opts *options) run() error {
...
+		if len(catalog) == 1 {
+			opts.from = catalog[0]
+		} else {
+			// Ask the catalog
+			err = selectOption.Ask("catalog", catalog)
+			if err != nil {
+				return err
+			}
+			opts.from = selectOption.Catalog
```
Comment (sm43) [inline api/pkg/cli/cmd/info/info.go:159]: can you move this to a seperate func

File: api/pkg/cli/cmd/info/info.go [hunk 12]
```diff
@@ -139,8 +139,70 @@ func (opts *options) run() error {
...
+		if err != nil {
+			return err
+		}
+
+		// Ask the resource name
+		err = selectOption.Ask(opts.kind, resources)
+		if err != nil {
+			return err
+		}
+		name = selectOption.Name
```
Comment (sm43) [inline api/pkg/cli/cmd/info/info.go:179]: same for this

File: api/pkg/cli/cmd/info/info.go (hunk 11 above)
Comment (PuneetPunamiya) [reply api/pkg/cli/cmd/info/info.go:159]: Any reason to move this into a seperate function ??

File: api/pkg/cli/hub/get_catalog.go [hunk 13]
```diff
@@ -0,0 +1,70 @@
+package hub
```
Comment (sm43) [inline api/pkg/cli/hub/get_catalog.go:1]: license

File: api/pkg/cli/cmd/info/info.go (hunk 11 above)
Comment (sm43) [reply api/pkg/cli/cmd/info/info.go:159]: readability :D

Comment (PuneetPunamiya) [reply api/pkg/cli/cmd/info/info.go]: Updated the pr!

File: api/pkg/cli/select_options/select_options.go [hunk 14]
```diff
@@ -0,0 +1,56 @@
...
+			},
+		},
+	}
+
+	if err := survey.Ask(qs, &ans, opts.AskOpts); err != nil {
+		return err
+	}
+
+	switch resourceInfo {
+	case "catalog":
```
Comment (pratap0007) [inline api/pkg/cli/select_options/select_options.go:47]: nit: can we  assign `catalog,task,version` to  variables , wdyt ?

**Comment by pratap0007** [comment]
WDYT, To mention the `latest` version while displaying all versions of a resource

File: api/go.mod [hunk 15]
```diff
@@ -3,21 +3,24 @@ module github.com/tektoncd/hub/api
...
 	github.com/ikawaha/goahttpcheck v1.3.1
 	github.com/joho/godotenv v1.3.0
 	github.com/mitchellh/go-homedir v1.1.0
 	github.com/sergi/go-diff v1.2.0 // indirect
-	github.com/spf13/cobra v1.1.1
+	github.com/spf13/cobra v1.1.3
 	github.com/spf13/viper v1.7.0
 	github.com/stretchr/testify v1.6.1
-	github.com/tektoncd/pipeline v0.20.1-0.20210204110343-8c5a751b53ea
+	github.com/tektoncd/pipeline v0.23.0
```
Comment (sm43) [inline api/go.mod:23]: can be upgraded to 0.24.1

File: api/pkg/cli/cmd/info/info.go [hunk 16]
```diff
@@ -138,9 +140,40 @@ func (opts *options) run() error {
...
+		if err != nil {
+			return err
+		}
+	}
+
+	name := opts.name()
+
+	// Check if resource name is passed else
+	// ask the user to select the resource
+	if name == "" {
```
Comment (sm43) [inline api/pkg/cli/cmd/info/info.go:159]: `opts.name == ""` ?

File: api/pkg/cli/cmd/info/info.go [hunk 17]
```diff
@@ -138,9 +140,40 @@ func (opts *options) run() error {
...
+		name, err = opts.AskResourceName()
```
Comment (sm43) [inline api/pkg/cli/cmd/info/info.go:160]: `opts.name, err = opts.AskResourceName()`

File: api/pkg/cli/cmd/info/info.go (hunk 16 above)
Comment (PuneetPunamiya) [reply api/pkg/cli/cmd/info/info.go:159]: `opts.name()` is a method

File: api/pkg/cli/cmd/info/info.go [hunk 18]
```diff
@@ -172,11 +205,79 @@ func (opts *options) run() error {
...
+		return "", err
+	}
+
+	// Ask the resource name
+	err = opts.selectOption.Ask(opts.kind, resources)
+	if err != nil {
+		return "", err
+	}
+	return opts.selectOption.Name, nil
+}
```
[the remark this answers is not in the document]
Comment (PuneetPunamiya) [reply api/pkg/cli/cmd/info/info.go:242]: It logs the message saying `No resources found`

[the remark this answers is not in the document]
Comment (vinamra28) [reply api/pkg/cli/cmd/info/info.go:242]: I see the error as 
image

File: api/pkg/cli/cmd/info/info.go [hunk 19]
```diff
@@ -172,11 +205,79 @@ func (opts *options) run() error {
 	return printer.New(out).Tabbed(tmpl, tmplData)
 }
 
+func (opts *options) AskCatalogName() (string, error) {
+	// Get all Catalogs
+	catalog, err := opts.hubClient.GetCatalogsList()
+	if err != nil {
+		return "", err
+	}
+	if len(catalog) == 1 {
```
Comment (piyush-garg) [inline api/pkg/cli/cmd/info/info.go:214]: handle if no catalog

File: api/pkg/cli/cmd/info/info.go [hunk 20]
```diff
@@ -172,11 +205,79 @@ func (opts *options) run() error {
...
+func (opts *options) AskResourceName() (string, error) {
+	// Get all resources from the Catalog selected
+	resources, err := opts.hubClient.GetResourcesList(hub.SearchOption{
+		Kinds:   []string{opts.kind},
+		Catalog: opts.from,
+	})
+	if err != nil {
+		return "", err
+	}
+
```
Comment (piyush-garg) [inline api/pkg/cli/cmd/info/info.go:235]: handle if no resource or 1 resource

File: api/pkg/cli/cmd/info/info.go [hunk 21]
```diff
@@ -172,11 +205,79 @@ func (opts *options) run() error {
...
+	// Get all the versions of the resource selected
+	ver, err := opts.hubClient.GetResourceVersionslist(hub.ResourceOption{
+		Name:    name,
+		Kind:    opts.kind,
+		Catalog: opts.from,
+	})
+	if err != nil {
+		return "", err
+	}
+	if len(ver) == 1 {
```
Comment (piyush-garg) [inline api/pkg/cli/cmd/info/info.go:254]: handle if no version

Comment (PuneetPunamiya) [reply api/pkg/cli/cmd/info/info.go:254]: If there is no version of the resource then it won't be a hub resource

Comment (PuneetPunamiya) [reply api/pkg/cli/cmd/info/info.go:254]: i.e. while parsing the catalog I guess the resource won't be added in the db

File: api/pkg/cli/cmd/info/info.go [hunk 22]
```diff
@@ -172,11 +206,88 @@ func (opts *options) run() error {
...
+func (opts *options) AskResourceName() (string, error) {
+	// Get all resources from the Catalog selected
+	resources, err := opts.hubClient.GetResourcesList(hub.SearchOption{
+		Kinds:   []string{opts.kind},
+		Catalog: opts.from,
+	})
+	if err != nil {
+		return "", err
+	}
+	if len(resources) == 0 {
```
Comment (sm43) [inline api/pkg/cli/cmd/info/info.go:239]: you can use `switch` here :)

File: api/pkg/cli/hub/get_catalog.go [hunk 23]
```diff
@@ -0,0 +1,84 @@
...
+	}
+
+	var data = struct {
+		Catalogs CatalogData
+	}{
+		Catalogs: typed,
+	}
+
+	Catalog := data.Catalogs
+	// // Get all catalog names
```
Comment (sm43) [inline api/pkg/cli/hub/get_catalog.go:77]: nit `// Get all catalog names`

File: api/pkg/cli/cmd/info/info.go (hunk 22 above)
Comment (PuneetPunamiya) [reply api/pkg/cli/cmd/info/info.go:239]: Will there be any more cases ??

File: api/pkg/cli/cmd/info/info.go [hunk 24]
```diff
@@ -172,11 +206,88 @@ func (opts *options) run() error {
 	return printer.New(out).Tabbed(tmpl, tmplData)
 }
 
+func (opts *options) AskCatalogName() (string, error) {
```
Comment (sm43) [inline api/pkg/cli/cmd/info/info.go:209]: Line no 200-207 won't get be executed now right?

File: api/pkg/cli/cmd/info/info.go (hunk 22 above)
Comment (sm43) [reply api/pkg/cli/cmd/info/info.go:239]: i would say yes :) [0,1,default]

File: api/pkg/cli/cmd/info/info.go (hunk 24 above)
Comment (PuneetPunamiya) [reply api/pkg/cli/cmd/info/info.go:209]: I think yes it won't be executed but in that case we don't need to even check if the version is empty, should we remove the condition to check if the version is empty ??

**Comment by pratap0007** [comment]
lgtm

**Comment by vinamra28** [comment]
/lgtm
Thanks :call_me_hand:
